Check that provider implements interface

Example from my code:

```
panic: reflect.Set: value of type *adapters.MockBankAdapter is not assignable to type processing.BankService

goroutine 1 [running]:
reflect.Value.assignTo(0x17eeb60, 0x20eb310, 0x16, 0x191d8be, 0xb, 0x181af60, 0xc0004fa1a0, 0x199, 0x20eb310, 0x1802c00)
	/usr/local/Cellar/go/1.12.5/libexec/src/reflect/value.go:2339 +0x437
reflect.Value.Set(0x181af60, 0xc0004fa1a0, 0x194, 0x17eeb60, 0x20eb310, 0x16)
	/usr/local/Cellar/go/1.12.5/libexec/src/reflect/value.go:1473 +0xa8
github.com/defval/inject/internal/graph.(*ProviderNode).Extract(0xc00037cc00, 0x181af60, 0xc0004fa1a0, 0x194, 0x181af60, 0xc00032f760)
	/Users/defval/go/pkg/mod/github.com/defval/inject@v1.3.1/internal/graph/node_provider.go:73 +0x2d0
github.com/defval/inject/internal/graph.(*InterfaceNode).Extract(0xc00033e960, 0x181af60, 0xc0004fa1a0, 0x194, 0xc0004fa1a0, 0x194)
	/Users/defval/go/pkg/mod/github.com/defval/inject@v1.3.1/internal/graph/node_interface.go:65 +0x1bb
github.com/defval/inject/internal/graph.(*ProviderNode).Extract(0xc00037c480, 0x1809900, 0xc000010390, 0x196, 0xc000010390, 0x196)
	/Users/defval/go/pkg/mod/github.com/defval/inject@v1.3.1/internal/graph/node_provider.go:55 +0x170
github.com/defval/inject/internal/graph.(*ProviderNode).Extract(0xc00037c980, 0x1814120, 0xc0004fa180, 0x194, 0x1814120, 0xc00032f9f8)
	/Users/defval/go/pkg/mod/github.com/defval/inject@v1.3.1/internal/graph/node_provider.go:55 +0x170
github.com/defval/inject/internal/graph.(*InterfaceNode).Extract(0xc00033e820, 0x1814120, 0xc0004fa180, 0x194, 0xc0004fa180, 0x194)
	/Users/defval/go/pkg/mod/github.com/defval/inject@v1.3.1/internal/graph/node_interface.go:65 +0x1bb
github.com/defval/inject/internal/graph.(*ProviderNode).Extract(0xc00037c300, 0x18b2040, 0xc000294008, 0x196, 0x18b2001, 0x17b1700)
	/Users/defval/go/pkg/mod/github.com/defval/inject@v1.3.1/internal/graph/node_provider.go:55 +0x170
github.com/defval/inject/internal/graph.(*Storage).Extract(0xc00000f0e0, 0x0, 0x0, 0x18b2040, 0xc000294008, 0x196, 0x20cf3a0, 0x2400d98)
	/Users/defval/go/pkg/mod/github.com/defval/inject@v1.3.1/internal/graph/storage.go:92 +0xf6
github.com/defval/inject.(*Container).Extract(0xc00008cdc0, 0x17798c0, 0xc000294008, 0x0, 0x0, 0x0, 0x0, 0xc)
	/Users/defval/go/pkg/mod/github.com/defval/inject@v1.3.1/container.go:72 +0x1e6
main.main()
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,9 +4,15 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## v1.4.2
+
+### Fixed
+
+- Replace: check that provider implement interface.
+
 ## v1.4.1
 
-## Fixed
+### Fixed
 
 - Lint
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,7 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## v1.4.0
 
-## Change
+### Change
 
 - `Container.WriteTo()` signature to `io.WriterTo`
 
@@ -36,7 +36,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## v1.2.1
 
-## Fixed
+### Fixed
 
 - inject.As() allows provide same interface without name
 

--- a/container_test.go
+++ b/container_test.go
@@ -543,7 +543,7 @@ func TestContainer_ProvideAs(t *testing.T) {
 			}, inject.As(new(http.Handler))),
 		)
 
-		require.EqualError(t, err, "could not compile container: type *net.TCPAddr not implement http.Handler interface")
+		require.EqualError(t, err, "could not compile container: could not create interface alias: type *net.TCPAddr not implement http.Handler interface")
 	})
 
 	t.Run("provide as interface with struct injection", func(t *testing.T) {
@@ -923,12 +923,21 @@ func TestContainer_Replace(t *testing.T) {
 		eqPtr(t, s, mockStringer)
 	})
 
-	t.Run("replace nil instanceProvider", func(t *testing.T) {
+	t.Run("replace nil provider", func(t *testing.T) {
 		_, err := inject.New(
 			inject.Replace(nil),
 		)
 
 		require.EqualError(t, err, "could not compile container: could not provide nil")
+	})
+
+	t.Run("replace not implemented interface", func(t *testing.T) {
+		_, err := inject.New(
+			inject.Provide(&net.TCPAddr{}, inject.As(new(net.Addr))),
+			inject.Replace(&http.Server{}, inject.As(new(net.Addr))),
+		)
+
+		require.EqualError(t, err, "could not compile container: could not create interface alias: type *http.Server not implement net.Addr interface")
 	})
 
 	t.Run("replace without interfaces", func(t *testing.T) {
@@ -946,7 +955,7 @@ func TestContainer_Replace(t *testing.T) {
 			inject.Replace(func() {}),
 		)
 
-		require.EqualError(t, err, "could not compile container: github.com/defval/inject_test.TestContainer_Replace.func5.1 constructor function must have at least one return value")
+		require.EqualError(t, err, "could not compile container: github.com/defval/inject_test.TestContainer_Replace.func6.1 constructor function must have at least one return value")
 	})
 
 	t.Run("replace already provided type", func(t *testing.T) {

--- a/internal/graph/node_interface.go
+++ b/internal/graph/node_interface.go
@@ -24,6 +24,10 @@ func NewInterfaceNode(name string, node *ProviderNode, iface interface{}) (_ *In
 		return nil, errors.Errorf("only interface supported") // todo: improve message
 	}
 
+	if !node.ResultType().Implements(typ) {
+		return nil, errors.Errorf("type %s not implement %s interface", node.ResultType(), typ)
+	}
+
 	return &InterfaceNode{
 		key: Key{
 			Type: typ,

